File: simulator/rvc_sim/_mock/_timer.py
from typing import Callable, Optional


class Timer:
    """Mirror of ``rvc::Timer`` (include/rvc/timer.hpp).

    Clock is injected as a callable returning monotonic milliseconds. Default
    clock returns 0 — callers in tests must inject a fake clock for any
    duration-dependent behavior.
    """

    def __init__(self, clock_fn: Optional[Callable[[], int]] = None) -> None:
        self._clock_fn: Callable[[], int] = clock_fn or (lambda: 0)
        self._callback: Optional[Callable[[], None]] = None
        self._duration_ms: int = 0
        self._start_time: int = 0
        self._is_running: bool = False
        self._is_fired: bool = False

    def setDuration(self, duration_ms: int) -> None:
        self._duration_ms = duration_ms

    def setCallback(self, callback: Callable[[], None]) -> None:
        self._callback = callback

    def start(self) -> None:
        self._start_time = self._clock_fn()
        self._is_running = True
        self._is_fired = False

    def stop(self) -> None:
        self._is_running = False

    def reset(self) -> None:
        self._is_running = False
        self._is_fired = False
        self._start_time = 0

    def isRunning(self) -> bool:
        return self._is_running

    def isExpired(self) -> bool:
        if not self._is_running:
            return False
        return (self._clock_fn() - self._start_time) >= self._duration_ms

    def update(self) -> None:
        if self._is_running and not self._is_fired and self.isExpired():
            self._is_fired = True
            self._is_running = False
            if self._callback is not None:
                self._callback()

File: src/timer.cpp
#include "rvc/timer.hpp"

#include <chrono>

namespace rvc {

Timer::Timer(ClockFn clock) : clock_{std::move(clock)} {
}

std::int64_t Timer::defaultClock() {
    return std::chrono::duration_cast<std::chrono::milliseconds>(
               std::chrono::steady_clock::now().time_since_epoch())
        .count();
}

void Timer::setDuration(std::int64_t durationMs) {
    duration_ = durationMs;
}

void Timer::setCallback(Callback callback) {
    callback_ = std::move(callback);
}

void Timer::start() {
    startTime_ = clock_();
    isRunning_ = true;
    isFired_ = false;
}

void Timer::stop() {
    isRunning_ = false;
}

void Timer::reset() {
    startTime_ = clock_();
    isFired_ = false;
}

bool Timer::isExpired() const {
    return isRunning_ && ((clock_() - startTime_) >= duration_);
}

bool Timer::isRunning() const {
    return isRunning_;
}

void Timer::update() {
    if (isRunning_ && !isFired_ && isExpired()) {
        isFired_ = true;

        if (callback_) {
            callback_();
        }
    }
}

} // namespace rvc